### PR TITLE
Bug - 4463 - Prevent application overwrite

### LINKED ISSUE #4463: API Allows users to overwrite applications
## 🐛 Bug

Currently, we only check if a user has already applied to a pool on the client during render. It is possible to have the page open in two different tabs/windows/computers/etc, apply for it in one, then in the separate location, apply again and overwrite the existing application.

h/t @petertgiles for finding this in https://github.com/GCTC-NTGC/gc-digital-talent/pull/4451#discussion_r1005953570

## 🎥 Recording

https://user-images.githubusercontent.com/4127998/198103866-751a9207-a890-4e1c-97b4-d93343664db0.mov

## ☔️ Possible Solution

We need update the mutation to throw the `ValidationException` earlier to prevent this and update the `CreateApplication.tsx` logic, moving it to the backend. 

```graphql
type Mutation {
  createApplication(userId: ID!, poolId: ID!): PoolCandidate
    @can(ability: "create", injectArgs: true)
}
```

```php
// App\GraphQL\Mutations\CreateApplication.php

$exists = PoolCandidate::where('user_id', $user->id)
  ->where('pool', $args['poolId'])
  ->exists();

$pool = Pool::find($args['poolId']);
if(in_array($pool->status, [POOL_ADVERTISEMENT_IS_DRAFT, POOL_ADVERTISEMENT_IS_EXPIRED])) {
  throw ValidationException::withMessages(['you are unable to apply to this pool']);
}

if($exists) {
  throw ValidationException::withMessages(['you have already applied to this pool']);
}

$application = PoolCandidate::create([
  'user_id' => $args['userId'],
  'pool_id' => $args['poolId'],
]);
```

```tsx
// CreateApplication.tsx

const isCreating =  !creating && !mutationData;
const hasRequiredData = userId && poolId;

if(!hasRequiredData) {
  redirect(redirectPath);
}

if(!isCreating) {
  executeMutation({ userId, poolId })
    .then((result) => {
      // Success, move onto application
    })
    .catch((err) => {
      // API throws error so we land here
    });
}

return <Loading />;
```

## ✅ Acceptance Criteria

- [ ] API does not allow users to recreate application
- [ ] API returns error if:
    - PoolCandidate already exists with `userId` and `poolId`
    - Pool is not published
 - [ ] `CreateApplication` logic simplified to handle fallback

### PR DESCRIPTION
**Pull request by esizer**
## 👋 Introduction

This rearranges the logic in the `CreateApplication` mutation to prevent users from overwriting existing applications. We now throw validation errors earlier. This helps to reduce the amount of business logic on the frontend and reduce odds of leaving users hanging.

## 🤯  Neat Stuff

GraphQL has the ability to return [partial errors](https://lighthouse-php.com/master/digging-deeper/error-handling.html#partial-errors). This PR uses that to return an `APPLICATION_EXISTS` along with the existing application so we can redirect a user to that existing application. 

## 🔨 WIP

- [x] Add new French strings once supplied

### Toast

On error, the toast is shown twice and using `toastId` is not preventing it for some reason. Is there a better way to prevent duplicate toasts?

## 🧪 Testing

1. Run PHPUnit tests
2. Reseed database
3. Build talent search `npm run production --workspace=talentsearch`
4. Login and navigate to the "Browse opportunities" page
5. Attempt to apply to a pool you have already applied to, you can do this by:
    1. Open the "Apply to this recruitment" link in two separate tabs
    2. Apply in the first tab
    3. Then, try to apply in the second tab
6. Ensure that the second application attempt fails

## 🤖 Robot Stuff

Resolves #4463

### CODE REVIEW

**Comment by petertgiles** [comment]
> do we have a mechnaism by which we can return codes to be translated appropriately on the frontend?

Yep, checkout https://github.com/GCTC-NTGC/gc-digital-talent/pull/4106

**Review by vd1992 — CHANGES_REQUESTED** [review]

File: api/app/GraphQL/Mutations/CreateApplication.php [hunk 1]
```diff
@@ -14,11 +16,28 @@ final class CreateApplication
...
+            throw ValidationException::withMessages(['You are unable to apply to this pool']);
+        }
+
+        // Do not allow users to apply if they already have
+        $exists = PoolCandidate::where('user_id', $args['userId'])
+            ->where('pool_id', $args['poolId'])
+            ->exists();
+        if ($exists) {
+            throw ValidationException::withMessages(['You have already applied to this pool']);
+        }
```
[the remark this answers is not in the document]
Comment (esizer) [reply api/app/GraphQL/Mutations/CreateApplication.php:34]: The problem was, that checks existence and it being after the creation was causing the issue of overwriting. The application still exists regardless if it has been submitted or not. We want the model to remain the same and let the user know they have already created an application. We don't really care if it has been submitted or not, just that it exists.

[the remark this answers is not in the document]
Comment (vd1992) [reply api/app/GraphQL/Mutations/CreateApplication.php:34]: I see, in that case the user probably ought to be shepherded towards /my-application, to either finish an application or view what they applied
which can be either toast or re-direct if you think it should be added

[the remark this answers is not in the document]
Comment (esizer) [reply api/app/GraphQL/Mutations/CreateApplication.php:34]: Yeah, that is the intention. This is still in progress and hasn't been completed yet.

File: frontend/common/src/messages/apiMessages.ts [hunk 2]
```diff
@@ -150,6 +150,12 @@ export const messages: { [key: string]: MessageDescriptor } = defineMessages({
     description:
       "Error message that the pool advertisement must have publishing group filled.",
   },
+  APPLICATION_EXISTS: {
+    defaultMessage: "You have already applied to this pool",
+    description:
+      "Message displayed when a user attempts to apply to pool more than once",
+    id: "0OPWbJ",
+  },
```
Comment (vd1992) [inline frontend/common/src/messages/apiMessages.ts:158]: need one for NOT_PUBLISHED

**Comment by vd1992** [comment]
clicking through it, it seems to work as intended! I don't follow entirely what is going on so I am going to waive my review

File: frontend/common/src/messages/apiMessages.ts (hunk 2 above)
Comment (esizer) [reply frontend/common/src/messages/apiMessages.ts:158]: Good eye, added in [add localized message for unpublished pool](https://github.com/GCTC-NTGC/gc-digital-talent/pull/4469/commits/3137b3feb8c7f13b3df316faea3b081376a88023)

**Review by petertgiles — APPROVED** [review]
I love everything about this PR, the rules and validators, the cool partial errors, the localized error messages.  :man_cook: :pinched_fingers:

File: api/app/GraphQL/Validators/Mutation/CreateApplicationValidator.php [hunk 3]
```diff
@@ -0,0 +1,36 @@
+<?php
+
+namespace App\GraphQL\Validators\Mutation;
+
+use App\Rules\NotAlreadyApplied;
+use App\Rules\PoolPublished;
+use Illuminate\Validation\Rule;
```
[the remark this answers is not in the document]
Comment (esizer) [reply api/app/GraphQL/Validators/Mutation/CreateApplicationValidator.php]: I either need a better colour theme to highlight these more prominently or we need a php linter 😆 

Removed in [remove unused ref](https://github.com/GCTC-NTGC/gc-digital-talent/pull/4469/commits/85c4dabc21ebb5f27188bdaa6f6838955af29a2f)

**Pull request merged**